File: 2º/PDOO/ProyectoRUBY/lib/mazo_sorpresas.rb
# encoding:utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
module Civitas
  class MazoSorpresas
    def initialize(bol='false')
      init
      @debug=bol
      @ultimaSorpresa
      if(@debug==true)
        e="Debug->true"
        ocurreEvento(e)
      end
    end
    def init()
      @sorpresas=[]
      @cartasEspeciales=[]
      @barajada=false
      @usadas=0
    end
    attr_reader :sorpresas
    attr_reader :cartasEspeciales
    attr_reader :barajada
    attr_reader :debug
    attr_reader :ultimaSorpresa
    attr_reader :usadas
    def alMazo(s)
      if(@barajada==false)
        @sorpresas<<s
      end
    end
    def siguiente
      if(@debug==true)
      else
        if(@barajada==false or @usadas==@sorpresas.size)
          @usadas=0
          @barajada=true
          @sorpresas.shuffle
        end
      end
      var=@usadas+1
      @usadas=var
      s=@sorpresas[0]
      @sorpresas.delete_at(0)
      @sorpresas<<s
      @ultimaSorpresa=s
      return s
    end
    def inhabilitarCartaEspecial(sorpresa)
      com=false
      carta=0
      tam=@sorpresas.size
      for i in 1..tam        
        if((@sorpresas[i]==sorpresa) and  (com==false))
          com=true
          carta=i
          @sorpresas.delete_at(carta)
          @cartasEspeciles<<sorpresa
          Diario.instance.ocurre_evento("Se elimina carta sorpresa #{sorpresa} del mazo y se añade a cartas especiales")
        end
      end
    end
    def habilitarCartaEspecial(sorpresa)
      com=false
      carta=0
      for i in 1..@cartasEspeciales.size
        if((@cartasEsciales[i]==@sorpresa) and  (com==false))
          com=true
          carta=i
          @cartasEspeciales.delete_at(carta)
          @sorpresas<<sorpresa
          a="Se elimina carta sorpresa #{sorpresa} de cartasEspeciales y se añade a mazo"
          ocurreEvento(a)
        end
      end
    end
    private :barajada, :debug, :ultimaSorpresa, :usadas
  end
end

File: 2º/PDOO/ProyectoJAVA/src/civitas/MazoSorpresas.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;
import java.util.Collections;

/**
 *
 * @author nikitastetskiy
 */
public class MazoSorpresas {
    
    private boolean barajada, debug;
    private int usadas;
    private Sorpresa ultimaSorpresa;
    
    private ArrayList<Sorpresa> sorpresas, cartasEspeciales; // cartasEspeciales como SALIRCARCEL
    
    void alMazo(Sorpresa s){
        if(!barajada){
            sorpresas.add(s);
        }
    }

    Sorpresa getUltimaSorpresa(){
        return ultimaSorpresa;
    }
    
    void habilitarCartaEspecial(Sorpresa sorpresa){
        if(cartasEspeciales.contains(sorpresa)){
            cartasEspeciales.remove(sorpresa);
            sorpresas.add(sorpresa);
            String e = "CartaEspecial removida y añadida a sorpresas";
            Diario.getInstance().ocurreEvento(e);
        }
    }
        
    void inhabilitarCartaEspecial(Sorpresa sorpresa){
        if(sorpresas.contains(sorpresa)){
            sorpresas.remove(sorpresa);
            cartasEspeciales.add(sorpresa);
            String e = "Carta sorpresa removida y añadida a cartasEspeciales";
            Diario.getInstance().ocurreEvento(e);
        }
    }
    
    private void init (){
        cartasEspeciales = new ArrayList<Sorpresa>();
        sorpresas = new ArrayList<Sorpresa>();
        ultimaSorpresa = null;                              // ?
        barajada = false;
        usadas = 0;
    }
    
    public MazoSorpresas(){
        init();
        debug = false;
    }
    
    public MazoSorpresas(boolean d){
        init();
        debug = d;
        if(debug){
            String e = "Debug-> true";
            Diario.getInstance().ocurreEvento(e);
        }
    }

    Sorpresa siguiente(){
        if((!barajada || usadas == sorpresas.size()) && !debug){
            usadas = 0;
            barajada = true;
            Collections.shuffle(sorpresas);
        }
        usadas ++;
        Sorpresa s = sorpresas.get(0);
        sorpresas.remove(0);
        sorpresas.add(s);
        ultimaSorpresa = s;
        return s;
    }

}
